### PR TITLE
feat: v25

### PR DESCRIPTION
**Pull request by rarkins**
Release v25. Individual commits within.

This should be merged using FF and not squash.

### CODE REVIEW

**Comment by rarkins** [comment]
I think this is now ready for merge, but will do it using rebase merge instead of squash, so that individual commits are preserved.

File: lib/manager/npm/post-update/lerna.ts [hunk 1]
```diff
@@ -96,10 +93,9 @@ export async function generateLockFiles(
       },
     };
     // istanbul ignore if
-    if (getAdminConfig().trustLevel === 'high') {
+    if (getAdminConfig().exposeAllEnv) {
       execOptions.extraEnv.NPM_AUTH = env.NPM_AUTH;
       execOptions.extraEnv.NPM_EMAIL = env.NPM_EMAIL;
-      execOptions.extraEnv.NPM_TOKEN = env.NPM_TOKEN;
     }
     if (config.dockerMapDotfiles) {
```
Comment (viceice) [inline lib/manager/npm/post-update/lerna.ts:100]: Wasn't this removed?

File: lib/manager/npm/post-update/lerna.ts [hunk 2]
```diff
@@ -96,10 +93,9 @@ export async function generateLockFiles(
       },
     };
     // istanbul ignore if
-    if (getAdminConfig().trustLevel === 'high') {
+    if (getAdminConfig().exposeAllEnv) {
       execOptions.extraEnv.NPM_AUTH = env.NPM_AUTH;
       execOptions.extraEnv.NPM_EMAIL = env.NPM_EMAIL;
-      execOptions.extraEnv.NPM_TOKEN = env.NPM_TOKEN;
```
Comment (viceice) [inline lib/manager/npm/post-update/lerna.ts:102]: Don't we pass whole env anyways?

**Review by viceice — COMMENTED** [review]
I thought we remove `dockerMapDotfiles` too with v25 🤔

**Comment by HonkingGoose** [comment]
@rarkins Does merging this PR auto-close related issues? I noticed we have some issues related to `.npmrc` and such that probably would be closed by landing this feature?

File: lib/config/definitions.ts [hunk 3]
```diff
@@ -461,17 +454,33 @@ const options: RenovateOptions[] = [
     default: false,
   },
   {
-    name: 'trustLevel',
+    name: 'exposeAllEnv',
     description:
-      'Set this to "high" if the bot should trust the repository owners/contents.',
+      'Configure this to true to allow passing of all env variables to package managers.',
```
Comment (HonkingGoose) [inline lib/config/definitions.ts:459]:
```suggestion
      'Configure this to true to allow passing of all environment variables to package managers.',
```

Maybe spell this out in full?

File: lib/datasource/crate/index.ts [hunk 4]
```diff
@@ -163,9 +163,9 @@ async function fetchRegistryInfo(
   };
 
   if (flavor !== RegistryFlavor.CratesIo) {
-    if (getAdminConfig().trustLevel !== 'high') {
+    if (!getAdminConfig().allowCustomCrateRegistries) {
       logger.warn(
-        'crate datasource: trustLevel=high is required for registries other than crates.io, bailing out'
+        'crate datasource: allowCustomCrateRegistries=true is required for registries other than crates.io, bailing out'
```
Comment (HonkingGoose) [inline lib/datasource/crate/index.ts:168]:
```suggestion
        'crate datasource: allowCustomCrateRegistries=true is required for registries other than crates.io, bailing out.'
```

End sentence with full stop.

File: lib/manager/terraform/readme.md [hunk 5]
```diff
@@ -1,21 +1,38 @@
...
+  - TerraformRegistry ( Public and Private )
+- providers ( deprecated in Terraform 0.13.0 )
+  - TerraformRegistry ( Public and Private )
+- required_providers block ( Terraform >= 0.13.0)
+  - TerraformRegistry ( Public and Private )
+- required_version
+- helm_release
+  - chart repository ( Public and Private )
+- docker\_\*
+  - Docker registry ( Public and Private )
```
Comment (HonkingGoose) [inline lib/manager/terraform/readme.md:15]: Would a table be clearer than the current list?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/manager/npm/post-update/lerna.ts (hunk 1 above)
Comment (rarkins) [reply lib/manager/npm/post-update/lerna.ts:100]: Should have been.. will add it now

File: lib/config/migration.ts [hunk 6]
```diff
@@ -189,12 +196,18 @@ export function migrateConfig(
...
+      } else if (key === 'trustLevel') {
+        delete migratedConfig.trustLevel;
+        if (val === 'high') {
+          migratedConfig.allowCustomCrateRegistries ??= true;
+          migratedConfig.allowScripts ??= true;
+          migratedConfig.exposeAllEnv ??= true;
         }
+      } else if (key === 'ignoreNpmrcFile') {
+        delete migratedConfig.ignoreNpmrcFile;
+        migratedConfig.npmrc ||= '';
```
Comment (JamieMagee) [inline lib/config/migration.ts:210]: What are the expected inputs here? `??=` works for `null` or `undefined`, whereas `||=` is for any 'falsy' values.

File: lib/manager/npm/post-update/lerna.ts (hunk 2 above)
Comment (rarkins) [reply lib/manager/npm/post-update/lerna.ts:102]: Not sure

File: lib/config/migration.ts (hunk 6 above)
Comment (rarkins) [reply lib/config/migration.ts]: Added explicit string check

File: lib/manager/terraform/readme.md (hunk 5 above)
Comment (rarkins) [reply lib/manager/terraform/readme.md:15]: Let's leave as-is. It's a tree concept so not a perfect fit for table either

**Comment by rarkins** [comment]
> @rarkins Does merging this PR auto-close related issues? I noticed we have some issues related to `.npmrc` and such that probably would be closed by landing this feature?

Yes, this should close #9111, #8792, #8291 and #2818

**Review by monstro805 — DISMISSED** [review]

File: lib/manager/npm/post-update/npm.ts [hunk 7]
```diff
@@ -49,7 +49,7 @@ export async function generateLockFile(
     const preCommands = [installNpm, 'hash -d npm'];
     const commands = [];
     let cmdOptions = '';
-    if (postUpdateOptions?.includes('npmDedupe') || skipInstalls === false) {
+    if (artifactUpdateApproach === 'deep') {
       logger.debug('Performing node_modules install');
       cmdOptions += '--ignore-scripts --no-audit';
```
Comment (viceice) [inline lib/manager/npm/post-update/npm.ts:54]: Ha, so no option to allow scripts. That's maybe a issue for some of our discussions about postinstall scripts.

needs a condition like `if (!getAdminConfig().allowScripts || config.ignoreScripts)` from `pnpm`

File: lib/manager/npm/extract/index.ts [hunk 8]
```diff
@@ -237,6 +237,10 @@ export async function extractPackageFile(
     }
     if (dep.currentValue.startsWith('file:')) {
       dep.skipReason = SkipReason.File;
+      // https://github.com/npm/cli/issues/1432
+      // Explanation:
+      //  - npm install --package-lock-only is buggy for transitive deps in file: references
+      //  - So we set artifactUpdateApproach to false if file: refs are found *and* the user hasn't explicitly set the value already
       hasFileRefs = true;
```
Comment (viceice) [inline lib/manager/npm/extract/index.ts:244]: https://github.com/renovatebot/renovate/commit/07affef41c6b4c7cb7d0413467800bdde4eb9876 is missing here

**Review by viceice — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**